src/multipart.rs

```diff
@@ -0,0 +1,172 @@
+use tokio::io::AsyncWrite;
+
+use crate::Result;
+
+type BoxedTryFuture<T> = Pin<Box<dyn Future<Output = Result<T, io::Error>> + Send>>;
+
+// Lifetimes are difficult to manage, so not using AsyncTrait
+pub(crate) trait CloudMultiPartUploadImpl {
+    /// Upload a single part
+    fn put_multipart_part(
+        &self,
+        buf: Vec<u8>,
+        part_idx: usize,
+    ) -> BoxFuture<'static, Result<(usize, UploadPart), io::Error>>;
+
+    /// Complete the upload with the provided parts
+    fn complete(
+        &self,
+        completed_parts: Vec<Option<UploadPart>>,
+    ) -> BoxFuture<'static, Result<(), io::Error>>;
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct UploadPart {
+    pub content_id: String,
+}
+
+pub(crate) struct CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl,
+{
+    inner: Arc<T>,
+    /// A list of completed parts, in sequential order.
+    completed_parts: Vec<Option<UploadPart>>,
+    /// Part upload tasks currently running
+    tasks: FuturesUnordered<BoxedTryFuture<(usize, UploadPart)>>,
+    /// Maximum number of upload tasks to run concurrently
+    max_concurrency: usize,
+    /// Buffer that will be sent in next upload.
+    current_buffer: Vec<u8>,
+    /// Minimum size of a part in bytes
+    min_part_size: usize,
+    /// Index of current part
+    current_part_idx: usize,
+    /// The completion task
+    completion_task: Option<BoxedTryFuture<()>>,
+}
+
+impl<T> CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl,
+{
+    pub fn new(inner: T, max_concurrency: usize) -> Self {
+        Self {
+            inner: Arc::new(inner),
+            completed_parts: Vec::new(),
+            tasks: FuturesUnordered::new(),
+            max_concurrency,
+            current_buffer: Vec::new(),
+            // TODO: Should self vary by provider?
+            // TODO: Should we automatically increase then when part index gets large?
+            min_part_size: 5_000_000,
+            current_part_idx: 0,
+            completion_task: None,
+        }
+    }
+
+    pub fn poll_tasks(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Result<(), io::Error> {
+        if self.tasks.is_empty() {
+            return Ok(());
+        }
+        let total_parts = self.completed_parts.len();
+        while let Poll::Ready(Some(res)) = self.tasks.poll_next_unpin(cx) {
+            let (part_idx, part) = res?;
+            self.completed_parts
+                .resize(std::cmp::max(part_idx + 1, total_parts), None);
+            self.completed_parts[part_idx] = Some(part);
+        }
+        Ok(())
+    }
+}
+
+impl<T> AsyncWrite for CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl + Send + Sync,
+{
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, io::Error>> {
+        // Poll current tasks
+        self.as_mut().poll_tasks(cx)?;
```

Comment:
I had no idea how this was working, turns out Rust 1.36 added some magic to make this work

https://doc.rust-lang.org/src/core/task/poll.rs.html#245